docs/CODE_REVIEW.md

```diff
@@ -0,0 +1,155 @@
+# How to do code reviews for curl
+
+Anyone and everyone is encouraged and welcome to review code submissions in
+curl. This is a guide on what to check for and how to perform a successful
+code review.
+
+## All submissions should get reviewed
+
+All pull requests and patches submitted to the project should be reviewed by
+at least one experienced curl maintainer before that code is accepted and
+merged.
```

Comment:
Is this also the case for core team members? I am currently thinking about this like:

If the change is simple, in line with existing things and has been tested by the core team member, the member can either directly push to master or create a PR for CI testing purposes, but it does not require a review.

Reply:
I went with "should" here as I believe this is what we *should* do. Persons with "merge access" in the project can always make the judgement call themselves if that is done or not, and I also didn't say that you can't review your own code. I believe **I** am one of the most prolific submit-PR-then-merge persons who land PRs without any second person having chimed in on it - and I believe I do that for the good of the project, not because I fear other people's comments - basically moving on when I think it is good enough after having given others a chance to comment/object/praise.

I still claim that *"at least one experienced curl maintainer should do a review before that code is accepted"*

Reply:
Okay, I am with you on this. I just did not think reviewing the PR yourself and giving others maybe a few hours to chime in fulfills this sentence, so maybe this could be made more explicit?

Reply:
I don't want to document shortcuts as it would sound like we encourage or suggest that to be a way we think is good.

I prefer to document how we *should* do and act. And then individuals with the given "authority" can decide to side-step the recommendations if they think it is beneficial for that particular case.

Reply:
> it would sound like we encourage or suggest that to be a way we think is good.

But doesn't this imply that not having at least one review is bad practice and discouraged?

Reply:
Yes!

Reply:
> And then individuals with the given "authority" can decide to side-step the recommendations if they think it is beneficial for that particular case.

Maybe we need a separate document describing what members are allowed to do. I would like to play by the rules which I am not sure I am up-to-date with.

Reply:
I think the [GOVERNACE](https://github.com/curl/curl/blob/master/docs/GOVERNANCE.md) document should be the one that talks about who is allowed to do what in the project in general.

Imagine a cookbook. It shows the recipe exactly how it should be done so that people know exactly how to do it, in particular if they never did these cookies before. But once you've baked those cookies many times, you might be allowed to take some liberties and freestyle a bit - for the sake of effectiveness and performance. That's wiggle room.

The cookbook doesn't document how or when to make shortcuts - and it rightfully doesn't even suggest that they are a good idea. That's something for every cook to do, or not to, according to your own judgement, skill and comfort-level. Going off-recipe is of course not recommended unless you really think you can still deliver at least the same quality end product. 

"freestyling" is done on your own responsibility and with the full awareness that it isn't by the book anymore and therefor I can rightfully be blamed if I burn down the house in the process.

I don't want to document that we're not allowed to burn down the house.

Reply:
Thanks for the detailed explanation!